src/pip/_internal/resolution/resolvelib/found_candidates.py

```diff
@@ -9,15 +9,31 @@
 """
 
 import functools
-from typing import Callable, Iterator, Optional, Set, Tuple
+from collections.abc import Sequence
+from typing import (
+    TYPE_CHECKING,
+    Callable,
+    Iterator,
+    Optional,
+    Set,
+    Tuple,
+    Union,
+    overload,
+)
 
 from pip._vendor.packaging.version import _BaseVersion
-from pip._vendor.six.moves import collections_abc  # type: ignore
 
 from .base import Candidate
 
 IndexCandidateInfo = Tuple[_BaseVersion, Callable[[], Optional[Candidate]]]
 
+if TYPE_CHECKING:
+    SequenceCandidate = Sequence[Candidate]
+else:
+    # Python 3.8
+    # TypeError: 'ABCMeta' object is not subscriptable
```

Comment:
It would be worthwhile to write this a bit clearer, e.g. *For compatibility: Python before 3.8 does not support using [] on class.*

Something like `TODO: Remove this block after dropping Python 3.7 support` would be useful as well.